images/virtualization-artifact/pkg/controller/vd/vd_webhook.go

```diff
@@ -0,0 +1,123 @@
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/cvicondition"
+)
+
+type Validator struct {
+	logger *slog.Logger
+}
+
+func NewValidator() *Validator {
+	return &Validator{
+		logger: slog.Default().With("controller", common.VDShortName, "webhook", "validator"),
+	}
+}
+
+func (v *Validator) ValidateCreate(_ context.Context, obj runtime.Object) (admission.Warnings, error) {
+	vd, ok := obj.(*virtv2.VirtualDisk)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualDisk but got a %T", obj)
+	}
+
+	v.logger.Info("Validating virtual disk", "spec.pvc.size", vd.Spec.PersistentVolumeClaim.Size)
+
+	if vd.Spec.PersistentVolumeClaim.Size != nil && vd.Spec.PersistentVolumeClaim.Size.IsZero() {
+		return nil, fmt.Errorf("virtual machine disk size must be greater than 0")
+	}
+
+	if vd.Spec.DataSource == nil && (vd.Spec.PersistentVolumeClaim.Size == nil || vd.Spec.PersistentVolumeClaim.Size.IsZero()) {
+		return nil, fmt.Errorf("if the data source is not specified, it's necessary to set spec.PersistentVolumeClaim.size to create blank virtual disk")
+	}
+
+	return nil, nil
+}
+
+func (v *Validator) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	newVD, ok := newObj.(*virtv2.VirtualDisk)
+	if !ok {
+		return nil, fmt.Errorf("expected a new VirtualDisk but got a %T", newObj)
+	}
+
+	oldVD, ok := oldObj.(*virtv2.VirtualDisk)
+	if !ok {
+		return nil, fmt.Errorf("expected an old VirtualDisk but got a %T", oldObj)
+	}
+
+	v.logger.Info("Validating virtual disk",
+		"old.spec.pvc.size", oldVD.Spec.PersistentVolumeClaim.Size,
+		"new.spec.pvc.size", newVD.Spec.PersistentVolumeClaim.Size,
+	)
+
+	if newVD.Spec.PersistentVolumeClaim.Size == oldVD.Spec.PersistentVolumeClaim.Size {
+		return nil, nil
+	}
+
+	if newVD.Spec.PersistentVolumeClaim.Size == nil {
+		return nil, errors.New("spec.persistentVolumeClaim.size cannot be omitted once set")
+	}
+
+	if newVD.Spec.PersistentVolumeClaim.Size.IsZero() {
+		return nil, fmt.Errorf("virtual machine disk size must be greater than 0")
+	}
+
+	if oldVD.Spec.PersistentVolumeClaim.Size != nil && newVD.Spec.PersistentVolumeClaim.Size.Cmp(*oldVD.Spec.PersistentVolumeClaim.Size) == -1 {
+		return nil, fmt.Errorf(
+			"spec.persistentVolumeClaim.size value (%s) should be greater than or equal to the current value (%s)",
+			newVD.Spec.PersistentVolumeClaim.Size.String(),
+			oldVD.Spec.PersistentVolumeClaim.Size.String(),
+		)
+	}
+
+	if oldVD.Generation == newVD.Generation {
+		return nil, nil
+	}
+
+	ready, _ := service.GetCondition(cvicondition.ReadyType, newVD.Status.Conditions)
+	if newVD.Status.Phase == virtv2.DiskReady || newVD.Status.Phase == virtv2.DiskLost || ready.Status == metav1.ConditionTrue {
+		if !reflect.DeepEqual(oldVD.Spec.DataSource, newVD.Spec.DataSource) {
+			return nil, fmt.Errorf("VirtualDisk is in a Ready state: data source cannot be changed")
```

Comment:
Message says "VirtualDisk is in a Ready state", but disk may be in DiskLost state.